src/main/java/com/proximyst/ban/service/impl/ImplMessageService.java

```diff
@@ -148,125 +118,140 @@ public ImplMessageService(final @NonNull IUserService userService,
+
+                "punishmentId", punishment.getId().orElse(-1L).toString(),
+                "punishmentDate", SimpleDateFormat.getDateInstance().format(punishment.getDate()),
+                "reason", punishment.getReason()
+                    .map(MiniMessage.get()::escapeTokens)
+                    .orElse("No reason specified"),
+                "punishmentType", punishment.getPunishmentType().name(),
+                "punishmentVerb", punishment.getPunishmentType().getVerbPastTense().map(this.cfg),
+
+                "expiry", !punishment.currentlyApplies()
+                    ? this.cfg.formatting.isLifted
+                    : punishment.isPermanent()
+                        ? this.cfg.formatting.never
+                        : DurationFormatUtils
+                            .formatDurationHMS(punishment.getExpiration() - System.currentTimeMillis()),
+                "duration", punishment.isPermanent()
+                    ? this.cfg.formatting.permanently
+                    : this.cfg.formatting.durationFormat
+                        .replace("<duration>",
+                            DurationFormatUtils.formatDurationWords(
+                                punishment.getExpiration() - System.currentTimeMillis(),
+                                false,
+                                false
+                            )))
+        )
+        .thenCompose(p -> this.formatMessage(messageKey, p));
   }
 
-  private @NonNull Component parseWithTarget(
-      final @NonNull String msg,
-      final @NonNull BanUser target) {
-    return MiniMessage.get().parse(
-        msg,
+  private @NonNull CompletableFuture<@NonNull Component> formatMessageFuture(final @NonNull String message,
+      final @Positive int futureCount, final @NonNull Object @NonNull ... placeholders) {
+    final CompletableFuture<?>[] futures = new CompletableFuture<?>[futureCount];
 
-        "targetName", target.getUsername(),
-        "targetUuid", target.getUuid().toString()
-    );
+    // We know how many futures there are, but we also need to actually find them.
+    int futureIdx = 0;
+    for (final Object placeholder : placeholders) {
+      if (placeholder instanceof CompletableFuture) {
+        futures[futureIdx++] = (CompletableFuture<?>) placeholder;
+      }
+    }
+
+    // We now need to know when all the futures are done.
+    // Luckily, we can do that with helper methods in CompletableFuture.
+    final CompletableFuture<?> allDone = CompletableFuture.allOf(futures);
+
+    return allDone.thenApply($ -> {
+      // We now know all the futures are done.
+      // We'll need to unwrap them, then pass to MiniMessage.
+      int found = 0;
+      for (int i = 0; i < placeholders.length; ++i) {
+        final Object obj = placeholders[i];
+        if (obj instanceof CompletableFuture) {
+          ++found;
+          placeholders[i] = String.valueOf(((CompletableFuture<?>) obj).join());
```

Comment:
More unsafe casts :(

Reply:
Also, wtf. This piece of code is not ok. The beautiful thing about strongly typed languages is type safety, yet you still managed to find a way to ruin this..

Reply:
Safe cast, look 2 lines up. This is needed because there will be uses where we need a future's value as a placeholder.